cmd/aws-s3-csi-controller/csicontroller/reconciler.go

```diff
@@ -0,0 +1,284 @@
+	}
+
+	return reconcile.Result{}, nil
+}
+
+// reconcileWorkloadPod reconciles given workload `pod` to spawn a Mountpoint Pod to provide a volume for it if needed.
+func (r *Reconciler) reconcileWorkloadPod(ctx context.Context, pod *corev1.Pod) (reconcile.Result, error) {
+	log := logf.FromContext(ctx).WithValues("pod", types.NamespacedName{Namespace: pod.Namespace, Name: pod.Name})
+
+	if pod.Status.Phase != corev1.PodPending {
+		log.V(debugLevel).Info("Pod is not pending - ignoring", "status", pod.Status.Phase)
+		return reconcile.Result{}, nil
+	}
+
+	if pod.Spec.NodeName == "" {
+		log.V(debugLevel).Info("Pod is not scheduled to a node yet - ignoring")
+		return reconcile.Result{}, nil
+	}
+
+	if len(pod.Spec.Volumes) == 0 {
+		log.V(debugLevel).Info("Pod has no volumes - ignoring")
+		return reconcile.Result{}, nil
+	}
+
+	var requeue bool
+	var errs []error
+
+	for _, vol := range pod.Spec.Volumes {
+		podPVC := vol.PersistentVolumeClaim
+		if podPVC == nil {
+			continue
+		}
+
+		// If PVC has no bound PVs yet, `getBoundPVForPodClaim` will return `errPVCIsNotBoundToAPV`.
+		// In this case we'll just return `reconcile.Result{Requeue: true}` here, which will bubble up to the
+		// original `Reconcile` call and will cause a retry for this Pod with an exponential backoff.
+		pvc, pv, err := r.getBoundPVForPodClaim(ctx, pod, podPVC)
+		if err != nil {
+			if errors.Is(err, errPVCIsNotBoundToAPV) {
+				requeue = true
+			} else {
+				errs = append(errs, err)
+			}
+			continue
+		}
+
+		csiSpec := extractCSISpecFromPV(pv)
+		if csiSpec == nil {
+			continue
+		}
+
+		log.V(debugLevel).Info("Found bound PV for PVC", "pvc", pvc.Name, "volumeName", pv.Name)
+
+		err = r.spawnMountpointPodIfNeeded(ctx, pod, pvc, pv, csiSpec)
+		if err != nil {
+			errs = append(errs, err)
+			continue
+		}
+	}
+
+	return reconcile.Result{Requeue: requeue}, errors.Join(errs...)
+}
+
+// spawnMountpointPodIfNeeded spawns a new Mountpoint Pod for given `pod` and volume
+// if there isn't already existing Mountpoint Pod.
+func (r *Reconciler) spawnMountpointPodIfNeeded(
+	ctx context.Context,
+	pod *corev1.Pod,
+	pvc *corev1.PersistentVolumeClaim,
+	pv *corev1.PersistentVolume,
+	csiSpec *corev1.CSIPersistentVolumeSource,
+) error {
+	mpPodName := mppod.MountpointPodNameFor(string(pod.UID), pvc.Spec.VolumeName)
+
+	log := logf.FromContext(ctx).WithValues(
+		"pod", types.NamespacedName{Namespace: pod.Namespace, Name: pod.Name},
+		"mountpointPod", mpPodName,
+		"pvc", pvc.Name, "volumeName", pv.Name)
+
+	err := r.Get(ctx, types.NamespacedName{Namespace: r.mountpointPodConfig.Namespace, Name: mpPodName}, &corev1.Pod{})
+	if err != nil && !apierrors.IsNotFound(err) {
+		log.Error(err, "Failed to get Mountpoint Pod")
+		return err
+	}
+
+	isMpExists := err == nil
+	if isMpExists {
+		log.V(debugLevel).Info("Mountpoint Pod already exists - ignoring")
+		return nil
+	}
+
+	if err := r.spawnMountpointPod(ctx, pod, pvc, pv, csiSpec, mpPodName); err != nil {
```

Comment:
Race condition on what? The `controller-runtime` and its underlying [`workqueue`](https://pkg.go.dev/k8s.io/client-go/util/workqueue) does not trigger concurrent reconciliation for the same key. See  `MaxConcurrentReconciles` on [`controller-runtime`](https://pkg.go.dev/sigs.k8s.io/controller-runtime/pkg/controller#TypedOptions) and `Stingy` property of [`workqueue`](https://pkg.go.dev/k8s.io/client-go/util/workqueue)

Reply:
If we ensure this can only be called once per MP instance at once we should be fine. Might be thorny later if we want to go for 1 MP instance for multiple pods